### ISSUE TITLE
about dice

### ISSUE THREAD

**Issue by 1164094277** [cwmok/Fast-Symmetric-Diffeomorphic-Image-Registration-with-Convolutional-Neural-Networks#8]
How do you calculate the dice? I add label image into Test_SYMNet.py and apply warped flow to get warped_label ,but the dice I caclulated is very low.
![1](https://user-images.githubusercontent.com/62985831/129034288-9475e035-e089-4ff6-abf0-21968b1617ee.jpg)
[1.zip](https://github.com/cwmok/Fast-Symmetric-Diffeomorphic-Image-Registration-with-Convolutional-Neural-Networks/files/6968607/1.zip)

**Comment by cwmok** [comment]
@1164094277 

It is not about the Dice score calculation.

This is the displacement field generated with my provided pre-trained model.
![mine_field](https://user-images.githubusercontent.com/63316551/129035925-dd06368b-f308-444e-a37a-e35a196abead.png)

This is the displacement field generated with your trained model.
![your_field](https://user-images.githubusercontent.com/63316551/129036056-bfff268e-6968-4ba8-a444-93c6a2c61bd7.png)

The deformation model used in SYMNet is called the stationary velocity field. The diffeomorphic properties only hold if and only if the velocity field is smooth. Apparently, the weight of smoothness regularization in your training setting is not sufficient to safeguard the smoothness's assumption.

Here is my advice:
1. Center-crop the data to match the data dimension with the one reported in our paper, and train the parameters mentioned in our paper (This avoid the complex parameter tuning process.)
OR
2. Set a higher weight for smoothness regularization (both smoothness regularization loss and orientation consistency loss). Then, add validation code to the training code (right after the code of model saving) to keep track of the model performance at different iterations, i.e. Dice score, std(Jacobian determinant) etc. You can stop the training once you observe that the smoothness of the deformation field is abnormal or the model coverage.

**Comment by 1164094277** [comment]
@cwmok @cwmok I've tried training to crop the image to the same size as your image, but the results aren&#039;t good either, as I mentioned in the previous question.(dice only 0.5369... ).
I've tried to set a higher weight , but the result is still poor.

**Comment by cwmok** [comment]
@1164094277 

Could you attach the trained model and training parameters for reference? 
Also, I want to clarify that we only used 26 anatomical structures for evaluation in our paper, including the Brain stem, thalamus, cerebellum cortex etc. But in your evaluation script, it includes 35 anatomical structures. 

Due to popular requests on the full-size OASIS dataset, I will upload the results as well as the pre-trained model on the full-size OASIS dataset next week.

Stay tuned.

**Comment by 1164094277** [comment]
@cwmok cwmok I checked the code and found that I was using "mode = ‘bilinear’ "(in SpatialTransform) when I applied the warpped flow to the label. I changed it to  "mode = ‘nearest' " to got better results（Pixel  in the warpped label is integer）.   I wonder if it‘s right .
![1](https://user-images.githubusercontent.com/62985831/129476935-5e17f0e8-e1c9-45ca-9563-6a88867d9f57.png)
![2](https://user-images.githubusercontent.com/62985831/129476936-21a5a922-9e9b-4325-ace5-8142de8945ed.png)

**Comment by cwmok** [comment]
@1164094277 

Yes, you are correct. I actually implemented two spatial transformation functions in `Models.py`: `SpatialTransform` and `SpatialTransformNearest`. The  `SpatialTransform` is for warping the image and the `SpatialTransformNearest` is for warping the label, so that you don't have to modify the `SpatialTransform`'s mode.

I am glad you have solved your problem.

**Comment by cwmok** [comment]
@1164094277 

I have added an example showing how to train on the preprocessed OASIS dataset. For details, please refer to the section "(Example) Training on the preprocessed OASIS dataset" in the readme file. Note that the resolution of the preprocessed dataset is too large for training. Therefore, we trained on the cropped data but tested on the full-resolution data. The results look good to me. For training details, you may look at the example in Train_sym_neurite_oasis.py and its data generator.
